maven/maven.bzl

```diff
@@ -89,20 +89,50 @@ def _convert_maven_dep(repo_name, artifact):
 def _normalize_target(full_target_spec, current_package, target_substitutions):
     full_target_spec = target_substitutions.get(full_target_spec, full_target_spec)
     full_package, target = full_target_spec.split(":")
-    local_package = full_package.split("//")[1] # @maven//blah/foo -> blah/foo
+    local_package = full_package.split("//")[1]  # @maven//blah/foo -> blah/foo
     if local_package == current_package:
-        return ":%s" % target # Trim to a local reference.
+        return ":%s" % target  # Trim to a local reference.
     return full_package if paths.filename(full_package) == target else full_target_spec
 
-def _get_dependencies_from_pom_files(ctx, artifact, group_path):
+# This should be in poms, but we want to keep ctx/network/file operations separate, and Starlark is constrained
+# enough that creating a "downloader" struct is more trouble than it's worth.
+def _fetch_pom(ctx, artifact):
     pom_urls = ["%s/%s" % (repo, artifact.pom) for repo in ctx.attr.repository_urls]
-    pom_file = "%s/%s-%s.pom" % (group_path, artifact.artifact_id, artifact.version)
+    pom_file = "%s/%s-%s.pom" % (artifact.group_path, artifact.artifact_id, artifact.version)
     ctx.download(url = pom_urls, output = pom_file)
-    result = ctx.execute(["cat" , pom_file])
+    result = ctx.execute(["cat", pom_file])
     if result.return_code:
-        fail("Error reading pom file %s (return code %s)", (pom_file, result.return_code))
+        fail("Error reading pom file %s (return code %s)" % (pom_file, result.return_code))
+    return result.stdout
+
+# In theory, this logic should live in poms.bzl, but bazel makes it harder to use the strategy pattern (to pass in a
+# downloader) and we want to keep file and network code out of the poms processing code.
+def _get_inheritance_chain(ctx, xml_text):
+    inheritance_chain = [poms.parse(xml_text)]
+    current = inheritance_chain[0]
+    for _ in range(100):  # Can't use recursion, so just iterate
```

Comment:
Is `while (true)` also disallowed?

Reply:
Also wondering if you should error out if you reach the end of the range.

Reply:
while(true) is disallowed.  There are only for-each loops over specific things - ranges, iterables, etc. 

I could error out, but I don't think there's a parent hierarchy greater than 100 in the whole maven corpus.